### DISCUSSION TITLE
App Router: Metadata dependent on User Agent

### DISCUSSION THREAD

**Discussion by fabb** [vercel/next.js#60661] [Help] [answered]
### Summary

I need to set the `viewport` to `"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no"` just for **iPhone** as a workaround for form field auto-zooming issues. 

### Additional information

With Pages router it was simple to just do a conditional `<Head>` element based on the user agent. In the App router it seems we don't have access to the user agent or other headers in the `generateViewport` function. How can we achieve this? It would be fine for me if it was overridden clientside-only after hydration. I thought about modifying the DOM in a useEffect, but I fear that this is a brittle solution and react could reset it on some next rerender.

**Comment by nirvikpurkait** [comment]
You can make use of metadata object in `app` router.

```jsx
export const metadata = {
	viewport: "",
};
```

**Reply by fabb** [reply]
no, as i wrote i need it dynamically based on user agent.

**Comment by nirvikpurkait** [comment, marked as answer]
Try using  `headers()` from `next/headers` to extract headers from a request, as the user-agent info is sent through request headers.

 Now you got the headers, from the headers try to conditionally check if the device is **_iPhone_** or not. 

```js
function isIphone(reqHeaders){
  // your code
}
```
`isIphone()` returns `boolean`

Now depending on the returned value you can chose what will be the value of `export const metadata = {
	viewport: "",
};`

**Reply by fabb** [reply]
thanks that works, but not with `export const metadata`  but rather `generateViewport`:

```
export async function generateViewport(): Promise<Viewport> {
    const userAgent = headers().get('user-agent')
    const isiPhone = /iphone/i.test(userAgent ?? '')

    return isiPhone
        ? {
              width: 'device-width',
              initialScale: 1,
              maximumScale: 1,
              userScalable: false,
          }
        : {}
}
```

**Reply by nickfujita** [reply]
This is awesome, but does this mean that server-side render caching is now different for every user-agent variation? (making cache hits less common)

**Reply by andrei-ce** [reply]
Thanks for the boilerplate @fabb! Gotta say I dislike this standard ios safari behavior at the very least haha.

Yes, anything that depends on the runtime needs to be in the generateViewPort function rather than the static ViewPost object: https://nextjs.org/docs/app/building-your-application/optimizing/metadata
